pkg/daytona/unmarshal.go

```diff
@@ -0,0 +1,251 @@
+package daytona
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+)
+
+const (
+	vaultTagPath = "vault_path"
+	vaultTagKey  = "vault_key"
+
+	vaultTagPathKey = "vault_path_key"
+
+	defaultVaultTagField = "value"
+)
+
+var ErrValueInput = errors.New("the provided value must be a struct pointer")
+
+type unmarshalSecretsConfig struct {
+	apex string
+}
+
+// An UnmarshalSecretsOption is an option for the UnmarshalSecrets function
+type UnmarshalSecretsOption interface {
+	apply(c *unmarshalSecretsConfig)
+}
+
+// WithApex returns an UnmarshalSecretsOption sets the apex value
+// for which the secrets unmarshaler should use
+func WithApex(apex string) UnmarshalSecretsOption {
+	return withApex(apex)
+}
+
+type withApex string
+
+func (w withApex) apply(c *unmarshalSecretsConfig) {
+	c.apex = string(w)
+}
+
+// UnmarshalSecrets traverses the value v recursively looking for tagged fields that
+// can be populated with secret data using the provided client and optional configured apex.
+// If the apex is configured, the tag vault_path_key is appended to the apex to construct
+// the final secret path. If the tag vault_path is provided, the apex is ignored.
+// A default key of 'default' is used on each path, it can be overridden using the vault_key tag.
+//
+//  Apex of 'secret/application' provided, combined to form 'secret/application/db_password'
+//  Field string `vault_path_key:"db_password"`
+//
+//  Apex of 'secret/application' provided, with key override
+//  Field string `vault_path_key:"db_password" vault_key:"password"`
+//
+//  vault_path represents a full secret path to fetch
+//  Field string `vault_path:"secret/application/db_password"`
+//  Field string `vault_path:"secret/application/db_password" vault_key:"password"` // key override
+func UnmarshalSecrets(client *api.Client, v interface{}, opts ...UnmarshalSecretsOption) error {
+	config := unmarshalSecretsConfig{}
+
+	for _, opt := range opts {
+		opt.apply(&config)
+	}
+
+	val := reflect.ValueOf(v)
+	if val.Kind() != reflect.Ptr {
+		return ErrValueInput
+	}
+
+	val = val.Elem()
+	if val.Kind() != reflect.Struct {
+		return ErrValueInput
+	}
+
+	for i := 0; i < val.NumField(); i++ {
+		fName := val.Type().Field(i).Name
+		f := val.Field(i)
+		if f.Kind() == reflect.Ptr {
+			if f.IsNil() {
+				f.Set(reflect.New(f.Type().Elem()))
+			}
+			f = f.Elem()
+		}
+
+		path, valueIndex := parsePath(config.apex, val.Type().Field(i).Tag)
+		switch f.Kind() {
+		case reflect.Struct:
+			err := UnmarshalSecrets(client, f.Addr().Interface(), WithApex(config.apex))
```

Comment:
instead of passing in `WithApex()` here we should just pass in `opts`, that _should_ make it more future proof if we add other options.

Reply:
This brings up user experience question regarding how the application would deal with more than one call to `UnmarshalSecrets`.  

The`WithApex()` option could be part of a top level struct receiver.  In the current implementation, the developer would have to provide `WithApex()` on every call to `UnmarshalSecrets` if there are multiple types to be unmarshalled. 

This could be adjusted to an experience similar to the following.

```go

// Create this once for the app, possibly changing the Apex based on environment during init
s := daytona.NewUnmarshaler(client).WithApex("secret/my/env/foo/super_secret_1")

// t1 fileld in from environment/config map/etc

// t2 fileld in from environment/config map/etc

// grab secrets for t1
s.UnmarshalSecrets(&t1)

// grab secrets for t2
s.UnmarshalSecrets(&t2)

```

I am still a fan of the variadic `opts` approach, but for something that may be "global" like an Apex  it may be nice to have that portion declared once.

Reply:
Or perhaps a situation where we need to crawl multiple apexes?
```
dayday := daytona.NewUnmarshaler(client)

app1 := dayday.WithApex("secret/app1/things")

app1.UnmarshalSecrets(&t1)
app1.UnmarshalSecrets(&t2)

...

app2 := dayday.WithApex("secret/app2/things")

app2.UnmarshalSecrets(&x)
app2.UnmarshalSecrets(&y)
```

Reply:
> instead of passing in `WithApex()` here we should just pass in `opts`, that _should_ make it more future proof if we add other options.

Good call. Resolved.